app/runner.go

```diff
@@ -0,0 +1,198 @@
+package app
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/grafana-app-sdk/logging"
+	"github.com/grafana/grafana-app-sdk/metrics"
+)
+
+var RunnableCollectorDefaultErrorHandler = func(ctx context.Context, err error) bool {
+	logging.FromContext(ctx).Error("runner exited with error", "error", err)
+	return true
+}
+
+// NewMultiRunner creates a new MultiRunner with Runners as an empty slice and ErrorHandler set to RunnableCollectorDefaultErrorHandler
+func NewMultiRunner() *MultiRunner {
+	return &MultiRunner{
+		Runners:      make([]Runnable, 0),
+		ErrorHandler: RunnableCollectorDefaultErrorHandler,
+	}
+}
+
+// MultiRunner implements Runnable for running multiple Runnable instances.
+type MultiRunner struct {
+	Runners []Runnable
+	// ErrorHandler is called if one of the Runners returns an error. If the function call returns true,
+	// the context will be canceled and all other Runners will also be prompted to exit.
+	// If ErrorHandler is nil, RunnableCollectorDefaultErrorHandler is used.
+	ErrorHandler func(context.Context, error) bool
+	// ExitWait is how long to wait for Runners to exit after ErrorHandler returns true or the context is canceled
+	// before stopping execution and returning a timeout error instead of exiting gracefully.
+	// If ExitWait is nil, Run execution will always block until all Runners have exited.
+	ExitWait *time.Duration
+}
+
+// Run runs all Runners in separate goroutines, and calls ErrorHandler if any of them exits early with an error.
+// If ErrorHandler returns true (or if there is no ErrorHandler), the other Runners are canceled and the error is returned.
+func (m *MultiRunner) Run(ctx context.Context) error {
+	propagatedContext, cancel := context.WithCancel(ctx)
```

Comment:
NIT – I'm not 100% sure but I think most of this can be replaced with https://pkg.go.dev/golang.org/x/sync/errgroup.

Reply:
> It looks like errgroup.Group will always wait for all goroutines to complete, so if we want to exit early on an error, we'd still need to track and close all the other ones.

It will indeed wait for all goroutines to complete, but it will cancel the internal context as soon as any goroutine returns an error, giving us the ability to either exit as soon as any goroutine returns an error OR the parent context is canceled. In both cases the benefit is that all underlying goroutines will be canceled early and ensured that they will run to completion.

The only difference would be that we will lose the ability to exit early without waiting for all goroutines to return, within a given timeout, but I'd argue this logic should only exist in the entrypoint of the app (i.e. the `main` function).

> We can't remove the errors channel in favor of it, either, because we need to be able to possibly respond differently based on the error.

Since `g.Wait` bubbles up the error from the underlying goroutine it already handles that for us, no?